src/dbus/error.cpp

```diff
@@ -0,0 +1,121 @@
+ */
+
+#include "dbus/error.hpp"
+
+#include "dbus/constants.hpp"
+#include "dbus/dbus_message_helper.hpp"
+
+#include <unordered_map>
+
+static const std::pair<otError, const char *> sErrorNameMap[] = {
+    {OT_ERROR_GENERIC, "io.openthread.Error.GenericError"},
+    {OT_ERROR_NONE, "io.openthread.Error.OK"},
+    {OT_ERROR_FAILED, "io.openthread.Error.Failed"},
+    {OT_ERROR_DROP, "io.openthread.Error.Drop"},
+    {OT_ERROR_NO_BUFS, "io.openthread.Error.NoBufs"},
+    {OT_ERROR_NO_ROUTE, "io.openthread.Error.NoRoute"},
+    {OT_ERROR_BUSY, "io.openthread.Error.Busy"},
+    {OT_ERROR_PARSE, "io.openthread.Error.Parse"},
+    {OT_ERROR_INVALID_ARGS, "io.openthread.Error.InvalidArgs"},
+    {OT_ERROR_SECURITY, "io.openthread.Error.Security"},
+    {OT_ERROR_ADDRESS_QUERY, "io.openthread.Error.AddressQuery"},
+    {OT_ERROR_NO_ADDRESS, "io.openthread.Error.NoAddress"},
+    {OT_ERROR_ABORT, "io.openthread.Error.Abort"},
+    {OT_ERROR_NOT_IMPLEMENTED, "io.openthread.Error.NotImplemented"},
+    {OT_ERROR_INVALID_STATE, "io.openthread.Error.InvalidState"},
+    {OT_ERROR_NO_ACK, "io.openthread.Error.NoAck"},
+    {OT_ERROR_CHANNEL_ACCESS_FAILURE, "io.openthread.Error.ChannelAccessFailure"},
+    {OT_ERROR_DETACHED, "io.openthread.Error.Detached"},
+    {OT_ERROR_FCS, "io.openthread.Error.FcsErr"},
+    {OT_ERROR_NO_FRAME_RECEIVED, "io.openthread.Error.NoFrameReceived"},
+    {OT_ERROR_UNKNOWN_NEIGHBOR, "io.openthread.Error.UnknownNeighbor"},
+    {OT_ERROR_INVALID_SOURCE_ADDRESS, "io.openthread.Error.InvalidSourceAddress"},
+    {OT_ERROR_ADDRESS_FILTERED, "io.openthread.Error.AddressFiltered"},
+    {OT_ERROR_DESTINATION_ADDRESS_FILTERED, "io.openthread.Error.DestinationAddressFiltered"},
+    {OT_ERROR_NOT_FOUND, "io.openthread.Error.NotFound"},
+    {OT_ERROR_ALREADY, "io.openthread.Error.Already"},
+    {OT_ERROR_IP6_ADDRESS_CREATION_FAILURE, "io.openthread.Error.Ipv6AddressCreationFailure"},
+    {OT_ERROR_NOT_CAPABLE, "io.openthread.Error.NotCapable"},
+    {OT_ERROR_RESPONSE_TIMEOUT, "io.openthread.Error.ResponseTimeout"},
+    {OT_ERROR_DUPLICATED, "io.openthread.Error.Duplicated"},
+    {OT_ERROR_REASSEMBLY_TIMEOUT, "io.openthread.Error.ReassemblyTimeout"},
+    {OT_ERROR_NOT_TMF, "io.openthread.Error.NotTmf"},
+    {OT_ERROR_NOT_LOWPAN_DATA_FRAME, "io.openthread.Error.NonLowpanDatatFrame"},
+    {OT_ERROR_LINK_MARGIN_LOW, "io.openthread.Error.LinkMarginLow"},
+};
+
+namespace otbr {
+namespace dbus {
+
+const char *ConvertToDBusErrorName(otError aError)
+{
+    const char *name = sErrorNameMap[0].second;
+
+    for (auto &&p : sErrorNameMap)
+    {
+        if (p.first == aError)
+        {
+            name = p.second;
+        }
+    }
+    return name;
+}
+
+otError ConvertFromDBusErrorName(const std::string &aErrorName)
+{
+    otError err = OT_ERROR_GENERIC;
+
+    for (auto &&p : sErrorNameMap)
+    {
+        if (p.second == aErrorName)
+        {
+            err = p.first;
+        }
+    }
+    return err;
+}
```

Comment:
Linear search is slow. Better to use 2 `unordered_map`s here.

Reply:
Maps, sets, and other dynamic containers: if you require a static, fixed collection, such as a set to search against or a lookup table, you cannot use the dynamic containers from the standard library as a static variable, since they have non-trivial destructors. Instead, consider a simple array of trivial types, e.g. an array of arrays of ints (for a "map from int to int"), or an array of pairs (e.g. pairs of int and const char*). For small collections, linear search is entirely sufficient (and efficient, due to memory locality); consider using the facilities from absl/algorithm/container.h for the standard operations. If necessary, keep the collection in sorted order and use a binary search algorithm. If you do really prefer a dynamic container from the standard library, consider using a function-local static pointer, as described below.